lxd/resources/pci.go

```diff
@@ -0,0 +1,118 @@
+package resources
+
+import (
+	"io/ioutil"
+	"path/filepath"
+	"strings"
+
+	"github.com/jaypipes/pcidb"
+	"github.com/pkg/errors"
+	"golang.org/x/sys/unix"
+
+	"github.com/lxc/lxd/shared/api"
+)
+
+// GetPCI returns a filled api.ResourcesPCI struct ready for use by LXD
+func GetPCI() (*api.ResourcesPCI, error) {
+	pci := api.ResourcesPCI{}
+
+	if !sysfsExists(sysBusPci) {
+		return &pci, nil
+	}
+
+	// Load PCI database
+	pciDB, err := pcidb.New()
+	if err != nil {
+		pciDB = nil
+	}
+
+	// Get uname for driver version
+	uname := unix.Utsname{}
+	err = unix.Uname(&uname)
+	if err != nil {
+		return nil, errors.Wrap(err, "Failed to get uname")
+	}
+
+	// List all PCI devices
+	entries, err := ioutil.ReadDir(sysBusPci)
+	if err != nil {
+		return nil, errors.Wrapf(err, "Failed to list %q", sysBusPci)
+	}
+
+	for _, entry := range entries {
+		entryName := entry.Name()
+		devicePath := filepath.Join(sysBusPci, entryName)
+		device := api.ResourcesPCIDevice{}
+
+		// Get driver name
+		driverPath := filepath.Join(devicePath, "driver")
+
+		if sysfsExists(driverPath) {
+			linkTarget, err := filepath.EvalSymlinks(driverPath)
+			if err != nil {
+				return nil, errors.Wrapf(err, "Failed to get driver of %q", devicePath)
+			}
+
+			device.Driver = filepath.Base(linkTarget)
+
+			// Try to get the version, fallback to kernel version
+			out, err := ioutil.ReadFile(filepath.Join(driverPath, "module", "version"))
+			if err == nil {
+				device.DriverVersion = strings.TrimSpace(string(out))
+			} else {
+				device.DriverVersion = strings.TrimRight(string(uname.Release[:]), "\x00")
+			}
+		}
+
+		// Get NUMA node
+		if sysfsExists(filepath.Join(devicePath, "numa_node")) {
+			numaNode, err := readInt(filepath.Join(devicePath, "numa_node"))
+			if err != nil {
+				return nil, errors.Wrapf(err, "Failed to read %q", filepath.Join(devicePath, "numa_node"))
+			}
+
+			if numaNode > 0 {
+				device.NUMANode = uint64(numaNode)
+			}
+		}
+
+		// Get PCI address
+		device.PCIAddress = entryName
+
+		// Get product ID node
+		productID, err := ioutil.ReadFile(filepath.Join(devicePath, "device"))
+		if err != nil {
+			return nil, errors.Wrapf(err, "Failed to read %q", filepath.Join(devicePath, "device"))
+		}
+
+		device.ProductID = strings.TrimPrefix(strings.TrimSpace(string(productID)), "0x")
+
+		// Get vendor ID node
+		vendorID, err := ioutil.ReadFile(filepath.Join(devicePath, "vendor"))
+		if err != nil {
+			return nil, errors.Wrapf(err, "Failed to read %q", filepath.Join(devicePath, "vendor"))
+		}
+
+		device.VendorID = strings.TrimPrefix(strings.TrimSpace(string(vendorID)), "0x")
```

Comment:
This doesn't match what we're doing so far and will fail if the files are missing.

Reply:
Fixed